### PR TITLE
Best management of TemplateType.

### PR DESCRIPTION
**Pull request by jDramaix**
Don't redefine TemplateType at method level if the same TemplateType is defined on the declaring Type.

### CODE REVIEW

**Comment by jDramaix** [comment]
fix for issue #277

File: src/main/main.iml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (rkirov) [inline src/main/main.iml]: revert this file

File: src/main/java/com/google/javascript/clutz/DeclarationGenerator.java [hunk 2]
```diff
@@ -1604,16 +1620,18 @@ private void aggregateFieldsFromClass(Set<String> fields, ObjectType superType)
         }
       }
       // visit prototype properties.
-      for (String field : superType.getConstructor().getPrototype().getOwnPropertyNames()) {
-        // getPropertyType works with non-owned property names, i.e. names from the prototype chain.
-        if (!superType.getPropertyType(field).isFunctionType()) {
-          fields.add(field);
+      if (superType.getConstructor() != null && superType.getConstructor().getPrototype() != null && superType.getConstructor().getPrototype().getOwnPropertyNames() != null) {
```
Comment (rkirov) [inline src/main/java/com/google/javascript/clutz/DeclarationGenerator.java]: line is too long, reformat

File: src/main/java/com/google/javascript/clutz/DeclarationGenerator.java [hunk 3]
```diff
@@ -1140,23 +1144,34 @@ private void visitVarDeclaration(String name, JSType type) {
...
+    private void visitTemplateTypes(ObjectType type, Set<String> alreadyEmittedTemplateType) {
       if (type.hasAnyTemplateTypes() && !type.getTemplateTypeMap().isEmpty()) {
-        List<String> templateKeys = new ArrayList<>();
-        for (TemplateType templateType: type.getTemplateTypeMap().getTemplateKeys()) {
+        List<String> realTemplateType = new ArrayList<>();
+
+        for (TemplateType templateType : type.getTemplateTypeMap().getTemplateKeys()) {
           String displayName = templateType.getDisplayName();
+
+          // Some TemplateType are already defined in a upper level (on Type definitions for
```
Comment (rkirov) [inline src/main/java/com/google/javascript/clutz/DeclarationGenerator.java]: The comment is a bit hard to read. Can you mention that this is a difference in semantics between Closure and TS. In closure having @template {T} declaration both at class-level and method-level make no difference, while in typescript, the a second template variable is introduced that shadows the first one.

File: src/main/java/com/google/javascript/clutz/DeclarationGenerator.java [hunk 4]
```diff
@@ -1626,24 +1644,43 @@ private void visitProperty(String propName, ObjectType objType, boolean isStatic
...
     private void emitProperty(String propName, JSType propertyType, boolean isStatic,
-                              boolean forcePropDeclaration) {
+        boolean forcePropDeclaration, Set<String> classTemplateTypeNames) {
       if (isStatic) emit("static");
       emit(propName);
       if (!propertyType.isFunctionType() || forcePropDeclaration) {
         visitTypeDeclaration(propertyType, false);
       } else {
-        visitFunctionDeclaration((FunctionType) propertyType);
+        // Avoid to re-emit templateType define on the Type level if method is not static.
```
Comment (rkirov) [inline src/main/java/com/google/javascript/clutz/DeclarationGenerator.java]: 'defined at the class-level'?

File: src/main/java/com/google/javascript/clutz/DeclarationGenerator.java [hunk 5]
```diff
@@ -1626,24 +1644,43 @@ private void visitProperty(String propName, ObjectType objType, boolean isStatic
...
+        visitFunctionDeclaration((FunctionType) propertyType, objTemplateTypes);
       }
       emit(";");
       emitBreak();
     }
 
-    private void visitFunctionDeclaration(FunctionType ftype) {
-      visitFunctionParameters(ftype);
+    private Set<String> getTemplateTypeNames(ObjectType objType) {
+      return objType.getTemplateTypeMap() != null
```
Comment (rkirov) [inline src/main/java/com/google/javascript/clutz/DeclarationGenerator.java]: too long for a ternary expression, can you reformat to an if statement.

File: src/main/java/com/google/javascript/clutz/DeclarationGenerator.java [hunk 6]
```diff
@@ -1663,12 +1700,13 @@ private void visitConstructorFunctionDeclaration(FunctionType ftype) {
     }
 
     private void visitFunctionParameters(FunctionType ftype) {
-      visitFunctionParameters(ftype, true);
+      visitFunctionParameters(ftype, true, Sets.<String>newHashSet());
```
Comment (rkirov) [inline src/main/java/com/google/javascript/clutz/DeclarationGenerator.java]: whats the difference between `Collections.<String>emptySet()` and `Sets.<String>newHashSet()`? I am no java expert, but use one of the two consistently.

**Comment by rkirov** [comment]
Looks good overall, just a few style fixes. This is quite a subtle difference between Closure and TS semantics, thanks for contributing the fix!

**Comment by jDramaix** [comment]
I've seen you committed the change yesterday. Do I have still to address your comments ?

**Comment by jDramaix** [comment]
I can do a follow-up CL if needed

**Comment by rkirov** [comment]
I pushed only the Index signature PR to master, but I will take care of this one today. No worries.

**Comment by rkirov** [comment]
took care of comments and merged.

**Pull request closed without merging**